packages/block-editor/src/components/url-input/index.js

```diff
@@ -18,233 +18,286 @@ import {
+	const [ prevInstanceId, setPrevInstanceId ] = useState( null );
+	const [ suggestionsListboxId, setSuggestionsListboxId ] = useState( '' );
+	const [ suggestionOptionIdPrefix, setSuggestionOptionIdPrefix ] = useState(
+		''
+	);
+	const [ prevDisableSuggestions, setPrevDisableSuggestions ] = useState(
+		null
+	);
+	const [ prevShowInitalSuggestions, setShowInitialSuggestions ] = useState(
+		null
+	);
+
+	const autocompleteRef = useRef( parentAutocompleteRef );
+	const inputRef = useRef( null );
+	const pendingRequest = useRef( null );
+	const scrollingIntoView = useRef( false );
+	const suggestionNodes = useRef( [] );
+	const instanceId = useInstanceId( URLInput );
+
+	const {
+		fetchLinkSuggestions = __experimentalFetchLinkSuggestions,
+	} = useSelect( ( select ) => {
+		if ( typeof __experimentalFetchLinkSuggestions === 'function' ) {
+			return {};
+		}
 
-			suggestionsListboxId: '',
-			suggestionOptionIdPrefix: '',
+		const { getSettings } = select( 'core/block-editor' );
+		const settings = getSettings();
+		return {
+			fetchLinkSuggestions: settings.__experimentalFetchLinkSuggestions,
 		};
-	}
+	} );
 
-	componentDidUpdate( prevProps ) {
-		const { showSuggestions, selectedSuggestion } = this.state;
-		const { value } = this.props;
-
-		// only have to worry about scrolling selected suggestion into view
-		// when already expanded
+	useEffect( () => {
+		// Only have to worry about scrolling selected suggestion into view when already expanded.
+		const selectedSuggestionNode =
+			suggestionNodes.current[ selectedSuggestion ];
 		if (
 			showSuggestions &&
-			selectedSuggestion !== null &&
-			this.suggestionNodes[ selectedSuggestion ] &&
-			! this.scrollingIntoView
+			!! selectedSuggestionNode &&
+			! scrollingIntoView.current
 		) {
-			this.scrollingIntoView = true;
-			scrollIntoView(
-				this.suggestionNodes[ selectedSuggestion ],
-				this.autocompleteRef.current,
-				{
-					onlyScrollIfNeeded: true,
-				}
-			);
+			scrollingIntoView.current = true;
+			scrollIntoView( selectedSuggestionNode, autocompleteRef.current, {
+				onlyScrollIfNeeded: true,
+			} );
 
-			this.props.setTimeout( () => {
-				this.scrollingIntoView = false;
+			setTimeout( () => {
+				scrollingIntoView.current = false;
 			}, 100 );
 		}
+	}, [ showSuggestions, value ] );
 
-		// Only attempt an update on suggestions if the input value has actually changed.
-		if (
-			prevProps.value !== value &&
-			this.shouldShowInitialSuggestions()
-		) {
-			this.updateSuggestions();
+	useEffect( () => {
+		const shouldShowInitialSuggestions =
+			! isUpdatingSuggestions() &&
+			showInitialSuggestions &&
+			! value?.length &&
+			! suggestions.length;
+
+		if ( shouldShowInitialSuggestions ) {
+			updateSuggestions();
 		}
+	}, [ value, showInitialSuggestions ] );
+
+	useEffect(
+		() =>
+			function cleanup() {
+				pendingRequest.current = null;
+			},
+		[]
+	);
+
+	if ( prevInstanceId !== instanceId ) {
```

Comment:
Good question, tbh I'm not sure since this logic is from the existing `getDerivedStateFromProps`. I tried following back through the commit history to see why `getDerivedStateFromProps` was being used but nothing jumped out. Then after some research on how to refactor `getDerivedStateFromProps` to use hooks I noticed that folks are just using comparing refs to state  outside of `useEffect`.